mmdet/utils/memory.py

```diff
@@ -0,0 +1,215 @@
+        2. When converting inputs to CPU, it will only look at each argument
+            and check if it has `.device` and `.to` for conversion. Nested
+            structures of tensors are not supported.
+        3. Since the function might be called more than once, it has to be
+            stateless.
+    """
+
+    def __init__(self, keep_type=True, convert_cpu=True, return_gpu=True):
+        self.logger = get_root_logger()
+        self.keep_type = keep_type
+        self.convert_cpu = convert_cpu
+        self.return_gpu = return_gpu
+
+    def retry_if_cuda_oom(self, func):
+        """Makes a function retry itself after encountering pytorch's CUDA OOM
+        error. It will do the following steps:
+
+            1. first retry after calling `torch.cuda.empty_cache()`.
+            2. If that still fails, it will then retry by trying to
+              convert inputs to FP16.
+            3. If that still fails trying to convert inputs to CPUs.
+              In this case, it expects the function to dispatch to
+              CPU implementation.
+
+        Args:
+            func: a stateless callable that takes tensor-like objects
+                as arguments
+        Returns:
+            func: a callable which retries `func` if OOM is encountered.
+        """
+
+        def maybe_to_cpu(x):
+            try:
+                like_gpu_tensor = x.device.type == 'cuda' and hasattr(x, 'to')
+            except AttributeError:
+                like_gpu_tensor = False
+            if like_gpu_tensor:
+                return x.to(device='cpu')
+            else:
+                return x
+
+        def maybe_to_fp16(x):
+            try:
+                like_float_tensor = \
+                    x.dtype == torch.float32 or x.dtype == torch.float64
+            except AttributeError:
+                like_float_tensor = False
+            if like_float_tensor:
+                return x.to(torch.half)
+            else:
+                return x
+
+        @wraps(func)
+        def wrapped(*args, **kwargs):
+
+            # raw function
+            with _ignore_torch_cuda_oom():
+                return func(*args, **kwargs)
+
+            # Clear cache and retry
+            torch.cuda.empty_cache()
+            with _ignore_torch_cuda_oom():
+                return func(*args, **kwargs)
+
+            # Try to use FP16
+            fp16_args = (maybe_to_fp16(x) for x in args)
+            fp16_kwargs = {k: maybe_to_fp16(v) for k, v in kwargs.items()}
+            self.logger.info(f'Attempting to copy inputs of {str(func)} '
+                             f'to fp16 due to CUDA OOM')
+
+            if self.keep_type:
+                # get input tensor type, the output type will same as
+                # the first parameter type.
+                if len(args) > 0:
+                    dtype = args[0].dtype
+                elif len(kwargs) > 0:
+                    dtype = list(kwargs.values())[0].dtype
```

Comment:
The first item of args and kwargs may not a tensor.

Reply:
Yes, you are right. But I did not have any idea how to avoid this potential error

Reply:
If we can iterate the args + kwargs.values(), find the first tensor and get its dtyoe and device.

Reply:
Ok